### ISSUE TITLE
uWSGI workers hang after respawn when using DDTrace Redis module with Redis Cluster

### ISSUE THREAD

**Issue by gamerdonkey** [DataDog/dd-trace-py#3072]
### Which version of dd-trace-py are you using?
`ddtrace==0.56.1`

### Which version of pip are you using?
`pip 21.1.1`

### Which version of the libraries are you using?
`Python 3.8.10`

```
attrs==21.2.0
backports.entry-points-selectable==1.1.1
boto3==1.9.102
botocore==1.12.130
build==0.7.0
certifi==2021.10.8
chardet==3.0.4
check-manifest==0.47
datadog==0.28.0
ddtrace==0.56.1
decorator==5.1.0
devpi-client==2.6.3
devpi-common==3.6.0
distlib==0.3.3
docker==3.6.0
docker-pycreds==0.4.0
docopt==0.6.2
docutils==0.18.1
es==2.24.0
filelock==3.4.0
gitdb2==3.0.0
gitlab-serialize==0.4.0
GitPython==2.1.11
idna==2.7
Jinja2==2.10
jmespath==0.10.0
lazy==1.4
MarkupSafe==2.0.1
packaging==20.3
pep517==0.12.0
pkginfo==1.8.2
platformdirs==2.4.0
pluggy==1.0.0
protobuf==3.19.1
py==1.11.0
pyparsing==3.0.6
python-dateutil==2.8.2
PyYAML==3.12
redis==3.5.3
redis-py-cluster==2.1.3
requests==2.20.1
s3transfer==0.2.1
sh==1.12.13
six==1.16.0
smmap==5.0.0
smmap2==3.0.1
tenacity==8.0.1
termcolor==1.1.0
toml==0.10.2
tomli==1.2.2
tox==3.24.4
urllib3==1.24.3
uWSGI==2.0.20
virtualenv==20.10.0
websocket-client==1.2.3
```

### How can we reproduce your problem?
Running a basic uWSGI application with tracing and a Redis cluster connection will allow a recreation of the issue.

#### main.py
```
from ddtrace import patch_all
from rediscluster import RedisCluster

print('Running DDTrace patch_all')
patch_all()

print('Making redis cluster connection.')
redis_cluser = RedisCluster(host='<redis cluster host>', port=6379, ssl=True, skip_full_coverage_check=True)


def application(env, start_response):
    start_response('200 OK', [('Content-Type','text/html')])
    return [b"Hello World"]
```
We modified the "Hello World" example to include DDTrace and RedisCluster. For our system, we are using AWS's MemoryDB, but any Redis cluster is most likely susceptible. The `ssl` and `skip_full_coverage_check` parameters allow connections to a default MemoryDB cluster.

#### requirements.txt
```
uwsgi==2.0.20
ddtrace==0.56.1
redis-py-cluster==2.1.3
```
Install with `pip install -r requirements.txt`.

#### run.sh
```
export UWSGI_MAX_REQUESTS=6

uwsgi --master-fifo /tmp/uwsgi.fifo --http :9090 --wsgi-file main.py
```
The **run.sh** script just sets up basic arguments and an environment variable that will force the uWSGI worker to respawn after 6 requests. By default there is only one worker.

After 6 requests to `localhost:9090`, the worker is respawned and a non-responsive worker will take its place. If `patch_all()` is commented out from **main.py**, the worker will respawn properly.

### What is the result that you get?
In production we run our application with multiple workers. They will all start up and begin serving requests as expected, but after a worker respawns for whatever reason (respawning is a normal operation in uWSGI) they become absolutely unresponsive. Eventually, all workers die in this way, and the entire application stops responding.

Initially there are no error messages, and uWSGI will continue trying to manage the workers. After some time (depending on application load), the request queue will fill up and we start seeing messages like this: `*** uWSGI listen queue of socket "0.0.0.0:5000" (fd: 6) full !!! (101/100) ***`

Attaching gdb to the stuck process in our application suggests that it's waiting on a mutex and that there may be some kind of deadlock with the GIL:

```
Attaching to process 958887
[New LWP 958888]
[New LWP 958889]
[New LWP 958890]
[New LWP 958891]
[New LWP 958892]
[New LWP 958893]
[New LWP 958894]
[New LWP 958895]
[New LWP 958896]
[New LWP 958897]
[New LWP 958898]
[New LWP 958899]
[New LWP 958900]
[New LWP 958901]
[New LWP 958902]
[New LWP 958903]
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
futex_wait_cancelable (private=<optimized out>, expected=0, futex_word=0x55f79e7a9788 <switch_cond+40>) at ../sysdeps/nptl/futex-internal.h:183
183    ../sysdeps/nptl/futex-internal.h: No such file or directory.
(gdb) list
178    in ../sysdeps/nptl/futex-internal.h
(gdb) where
#0  futex_wait_cancelable (private=<optimized out>, expected=0, futex_word=0x55f79e7a9788 <switch_cond+40>) at ../sysdeps/nptl/futex-internal.h:183
#1  __pthread_cond_wait_common (abstime=0x0, clockid=0, mutex=0x55f79e7a9720 <switch_mutex>, cond=0x55f79e7a9760 <switch_cond>) at pthread_cond_wait.c:508
#2  __pthread_cond_wait (cond=cond@entry=0x55f79e7a9760 <switch_cond>, mutex=mutex@entry=0x55f79e7a9720 <switch_mutex>) at pthread_cond_wait.c:638
#3  0x000055f79e54a49f in drop_gil (tstate=tstate@entry=0x55f79f108f90) at Python/ceval_gil.h:200
#4  0x000055f79e552330 in _PyEval_EvalFrameDefault (f=<optimized out>, throwflag=<optimized out>) at Python/ceval.c:1190
#5  0x000055f79e54b999 in PyEval_EvalFrameEx (throwflag=0, f=0x7fbb86660428) at Python/ceval.c:754
#6  _PyEval_EvalCodeWithName (_co=_co@entry=0x7fbb86583030, globals=globals@entry=0x7fbb86612090, locals=locals@entry=0x0, args=args@entry=0x7ffc2471aae0, 
    argcount=argcount@entry=2, kwnames=kwnames@entry=0x0, kwargs=0x0, kwcount=0, kwstep=2, defs=0x7fbb8660efa0, defcount=2, kwdefs=0x0, 
    closure=0x7fbb86587cf8, name=0x7fbb8664c180, qualname=0x7fbb865822b0) at Python/ceval.c:4166
#7  0x000055f79e5552a9 in _PyFunction_FastCallDict (func=func@entry=0x7fbb8661da60, args=args@entry=0x7ffc2471aae0, nargs=nargs@entry=2, 
    kwargs=kwargs@entry=0x0) at Python/ceval.c:5084
#8  0x000055f79e48cf59 in _PyObject_FastCallDict (func=0x7fbb8661da60, args=0x7ffc2471aae0, nargs=2, kwargs=0x0) at Objects/abstract.c:2310
#9  0x000055f79e48d0a5 in _PyObject_Call_Prepend (func=0x7fbb8661da60, obj=<optimized out>, args=0x7fbb867bea20, kwargs=0x0) at Objects/abstract.c:2373
#10 0x000055f79e48ccee in PyObject_Call (func=0x7fbb868fd488, args=<optimized out>, kwargs=<optimized out>) at Objects/abstract.c:2261
#11 0x000055f79e456d0b in uwsgi_python_reset_random_seed ()
#12 0x000055f79e45865c in uwsgi_python_post_fork ()
#13 0x000055f79e44596e in uwsgi_run ()
#14 0x000055f79e3efa54 in main ()
```

### What is the result that you expected?
Multi-threaded workers in uWSGI should be able to respawn successfully with DDTrace patching and Redis cluster connections. Normal, non-clustered Redis connections still seem to work without issue.

We've been able to work around this problem by disabling Redis tracing with `patch_all(redis=False)` or the environment variable `DD_TRACE_REDIS_ENABLED=false`.

Thanks for looking at this! Let us know if you need any more information.

**Comment by brettlangdon** [comment]
@gamerdonkey would you be able to try with: `--enable-threads` and `--lazy-apps`?

Based on our docs here: https://ddtrace.readthedocs.io/en/stable/advanced_usage.html#uwsgi

The really short version of the story is, there is a known issue with `uWSGI` where if a Python thread gets started in the main process then there is a high likelihood of worker threads failing to respawn due to threadstate/lock cleanup ordering post fork. This is an issue whether you are using our library our not, but since we start background threads there is a chance, depending on usage, that we are the ones triggering the bad behavior.

For now we ask people to ensure that you set:
- `--enable-threads` - we just need this to run, since we start threads
- `--lazy-apps` - delays loading the app in the main process, instead loads it in the worker process
- `--import=ddtrace.bootstrap.sitecustomize` - _optional_ but required if you wanted the same benefits of `ddtrace-run` (automatically calling `patch_all()` for you, etc)


I am unsure why disabling redis patching resolves the issue, so I am hesitant to say that my above description/suggestion will resolve it 100%, but the symptoms sound similar enough it would be worth trying.

**Comment by majorgreys** [comment]
Thanks for the detailed issue report @gamerdonkey!

I wanted to make sure that there are no other uwsgi configuration options than the one passed with the exec command you shared. I don't see an ini file.

**Comment by majorgreys** [comment]
Also can you share the log from uWSGI where it starts up and fails to respawn?

**Comment by majorgreys** [comment]
@gamerdonkey I wonder if it might not be because you are not enabling threads support in uWSGI. I was able to reproduce the problem when workers respawn given what you shared. If I enable threads support, the worker does not hang. Can you confirm if that addresses the problem you are experiencing?

https://gist.github.com/majorgreys/a65cd09f04c819662edab0113e1d5134

**Comment by gamerdonkey** [comment]
@brettlangdon @majorgreys  Thank you for the quick responses and the documentation links. I'm honestly amazed we haven't come across that page before now in all the research we've done.

For an update, today we were able to test with `--enable-threads` and `--lazy-apps` and it did clear up the hanging workers problem. It also resulted in a pretty significant increase in resource usage and some performance hits. We're still investigating that, but we'll have to examine this impact this might have before setting those options in our production environment.

@majorgreys In answer to your question, our production application does set a number of options in its ini, but for the recreation example I found that default options worked so I just went with no ini. Also, as you might have seen by now, the logs aren't very interesting. 😝 No errors from the failing workers.

It's very interesting that just using the `enable-threads` option was enough to let the workers respawn. We'll look into if that could help with some of the performance effects we've been seeing.

I really like the environment you set up for testing. Thanks for sharing the gist.
